merge grassland and soil land use proceses into (more) latest version

jnnsbrr opened the pull request.

chrisdwells commented:
Thanks Jannes! I started making notes on this before I realised it was a draft version, so thought may as well send over anyway though I'm sure you're aware of the below- cheers!

--

Undefined units; think it'd be good to get those in so it's more readily understandable to the group. 

In "soil carbon decay" there's "initial surface temperature anomaly", which is a set value and the units aren't degC. With impacts generally we've been locking things to pre-industrial levels so there's variation between members in 1980 and to make it less sensitive to the 1980 temperature anomaly. If there's good reason to change from that for some impacts then we definitely can, but good to chat through if that's OK. Those units aren't right for the temperature anomaly, and we should be careful as that anomaly (ie 1980 value) is dependent on the run. Also in this module, the decay rates are 1/year for fast and slow, but dmnl for litter - should all be 1/year?

In "cropland soil carbon" there's "soil carbon emissions cropland GtC" which is forced to be positive; it hits zero in a lot of the run - what's the rationale for not going negative?

Just in terms of the naming, there's a "soil carbon emissions" module which has soil carbon emissions and litter emissions - should the module or one of those be renamed to emphasise the sum of soil and letter (unless I'm misunderstanding the process)?

jnnsbrr commented:
Dear @wasbridge,
on the issues you mentioned:
* I added units (I guess almost everywhere)
* Together with @chrisdwells I added a fix for initial surface temperature in the climate module
* Now also emissions could also be negative
* For the soil carbon emissions module - we can rename it, I am quite flexible here.

I still have one big problem which I tried to solve the whole day but failed ...
I added a terrestrial carbon balance module which is a check that I am not missing any carbon throughout the system.
`Terrestrial carbon balance over time` should be zero, but it is not. It accumulates carbon which I must have missed somewhere but I cannot find the reason ... maybe you have a clue?

This is the major problem left, everything else should be almost ready to merge!

Thanks in advance :-)

wasbridge commented:
I'm on vacation till Tuesday.  I'll look then.  My first instinct is a
missing land use change flow.  Thanks for all the hard work!

On Thu, Sep 19, 2024, 12:10 PM Jannes Breier ***@***.***>
wrote:

> Dear @wasbridge <https://github.com/wasbridge>,
> on the issues you mentioned:
>
>    - I added units (I guess almost everywhere)
>    - Together with @chrisdwells <https://github.com/chrisdwells> I added
>    a fix for initial surface temperature in the climate module
>    - Now also emissions could also be negative
>    - For the soil carbon emissions module - we can rename it, I am quite
>    flexible here.
>
> I still have one big problem which I tried to solve the whole day but
> failed ...
> I added a terrestrial carbon balance module which is a check that I am not
> missing any carbon throughout the system.
> Terrestrial carbon balance over time should be zero, but it is not. It
> accumulates carbon which I must have missed somewhere but I cannot find the
> reason ... maybe you have a clue?
>
> This is the major problem left, everything else should be almost ready to
> merge!
>
> Thanks in advance :-)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/metno/WorldTransFRIDA/pull/34#issuecomment-2361440838>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGMI2VOVGPNTM24O7NXZPDZXLZRBAVCNFSM6AAAAABOCXZQVSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNRRGQ2DAOBTHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

chrisdwells commented:
Jannes I'd say first thing is to sort out the units; needs doing anyway and that might show up the issue. Just conceptually though I'm not sure I see how the balance works, though it's completely not my field (ha ha). Would we expect the soil emissions to balance at the same timestep as the production? The emissions are driven by the carbon stocks, so don't directly match up with the production inputs?

jnnsbrr commented:
Dear @chrisdwells,
thanks a lot for the first quick review!
I fixed the missing units.
I implemented the balance as a test to make sure no carbon gets lost along the way or no carbon is introduced out of nowhere into the system (these kind of tests are a sort of standard in larger land (carbon) cycle models. Still I am not sure if the test is valid in the way it has been implemented in a SD way. @wasbridge would be really helpful to talk to you on that next Tuesday. I think with this amount of carbon processes we have to make sure the model/module is consistent. Its easy to mess it up 😬

lnnrtrmm commented:
Dear @jnnsbrr,

I had a look at the implementation this morning, but I also couldn't find the error in the accounting.

There is, however, another issue that worries me even more. The sum of the carbon stocks actually looses carbon over time (~24 GtC between 1980 and 2024), which is the opposite of what it should do, right? The land carbon sink took up ~25% of annual total co2 emissios during the last decades, whereas the land-use change emissions make up 10% of all emissions currently (20% in 1980). Hence, there should be a net increase of the total land carbon pool, which we don't have here. Or is this because all this is not calibrated yet?

jnnsbrr commented:
Dear @lnnrtrmm thanks a lot for your review!
And good though that you also did not find any major issues.
Yes right - that is what we meant. I am still not sure if we are missing something or if the approach itself is not valid. I really hope Billy can shed some light on the matter 😬

wasbridge commented:
@jnnsbrr So I just fixed all the unit errors, and got the existing structure roughly in line with history again.  With that fixed its easier to look at the problem.

The thing that sticks out to me here is the spikes from CO2 in the future in Net Primary Production are TOTALLY missing from "change in stocks".  That points to just one source of the missing balance... those CO2 effects impact net primary production, and terrestrial carbon loss, but not soil carbon balance cropland, soil carbon balance grassland etc etc.

What I would do is take what I have commited and in your Terrestrial carbon balance module make balances for each land use type.  i.e. for cropland I would have a cumulative terrestrial cropland carbon balance, which adds in the difference between cropland net primary production and crop production in GTC and all of the other changes (soil, production, residue etc).  By looking at each land type individually you should be able to find out either what things you're missing in your balance calculation, or what things are making the balance go wacky.  Going land use type by land use type should also help identify the issue raised by @lnnrtrmm

jnnsbrr commented:
I fixed it!!
Thanks a lot @wasbridge for the advices. I added a check for each land use type and found out that it was the forest where I made a mistake in accounting the forest soil carbon emissions. Now every carbon balance equation should yield in 0 :-)
@wasbridge is there a built in to throw an error if its not ~0? And if so we should add it I guess ...

lnnrtrmm commented:
Thanks @jnnsbrr, great that you fixed the balance!

The overall issue I was mentioning still remains: The land looses 150 GT C between 1980 and 2100, whereas it should actually take up carbon in response to increasing atmospheric CO2. I am not sure, whether this is because of some missing process or because we are deforesting the land way too much. I assume deforestation etc. are calibrated in the initial period, so the fact that already during that period we see a loss of land carbon (which is not in line with historical data) tells me that we are missing a process that leads to increased land carbon storage under elevated CO2.

I don't have any time left for today, so I'll take a closer look tomorrow. But I just want to mention that if we would couple this as it is right now with the overall carbon cycle (including also an ocean carbon cycle), we would get a huge increase of atmospheric CO2, because we are essentially missing the land carbon sink.

jnnsbrr commented:
Dear @lnnrtrmm,
the carbon source is due to land use change emissions. We are shifting a lot of land around and these are the consequences. I just did a run where I set land use change to zero and tadaa we have a sink again ...
![no-landuse](https://github.com/user-attachments/assets/4136924e-b0a3-43b2-8bdd-c646591c3776)
At least until ~2060, then its decreasing again because grazing goes up insanely high, decay is also increased with temperature and some pools like carbon tend to have slow response because of the long mean residence time of the slow soil carbon pool.

lnnrtrmm commented:
Thanks for the clarification, so it was indeed the fact that land-use change was not calibrated well, and it does look more like I would expect, when you deactivated land-use change, that is nice!

I would just like to point out that even without land-use change, the land carbon sink seems to be a factor 2-3 too small (although this might partly be a consequence of the whole system being out of line with calibration data now), so potentially not just land-use change, but also the land carbon sink itself need further calibration. We can take the calibration data for the land carbon sink from the GCB, just like you already have for CO2 emissions from land use change (https://globalcarbonbudget.org/download/933/?tmstv=1701441499).
The net annual land carbon uptake in FRIDA should then be equal to the difference between land carbon sink and land-use change emissions in the GCB data.

Also, I apologize for having so many comments... It is actually a great achievement and step forward that we have a mass conserving accounting of the land carbon cycle in FRIDA! Not many models have that.

wasbridge commented:
@jnnsbrr Lets confer Friday as well about how to get the latest F&L merged into your branch, so that you can be working with the latest and greatest, so that once the land sink is sorted its a painless merge.  I can likely do the technical aspects assuming you have no non-pushed commits in your fork

wasbridge commented:
@jnnsbrr @lnnrtrmm taking the average_time_cropland_in_use to 1000 or 10000 stops all the land cycling (without affecting the stock values) and makes the "change_in_stocks" in the terrestial carbon module be positive for a while in the beginning, so I think that is half the problem.  To change depleted land change "fraction cropland degrading 1980",  "fraction grassland degrading 1980",    "elasticity of effect of crop yield on fraction cropland degrading" and the equiv variable for grassland.

jnnsbrr commented:
Dear @wasbridge thanks a lot, yes lets have a closer look today.

wasbridge commented:
@jnnsbrr this branch is now based on mainline.  Go look at soil carbon emissions in the Emissions module.  I disconnected them b/c they appear to be an order of magnitude too high.  Putting those crazy high emissions into Food and Land Use CO2 emissions changes the terrestrial carbon balance somewhat signficantly

jnnsbrr commented:
Dear @wasbridge, yes that has to be rethought as well. For now we have to use the net soil carbon emissions, the difference of soil in and outputs.

wasbridge commented:
@jnnsbrr I put those emissions back in now

lnnrtrmm commented:
Dear @jnnsbrr @wasbridge,

I just pushed some calibration into jannes main branch, but these are only for the global land carbon balance and forest clearing / regrowth from GCB. For grassland and cropland there were no fitting data in the GCB at least.

However, I found another issue that could be one of the problems we have: Young forests are loosing a lot of carbon in the initial decades and this is because their carbon stocks are way too large. For comparison: The young forest area has just 0.5 % of all the forest area in 1980, but the young forest carbon stocks account for ~16% of all forest carbon. This means young forest initially have > 30x as much carbon per area than mature forests in 1980, but they should actually have substantially less carbon per area! Because of this the whole system is initially out of equilibrium and the young forests loose much of their carbon in the first years, substantially affecting our carbon balance. To fix this, we should either put a larger fraction of the area into young forests, or move some of the initial carbon from young forest to mature forest.

lnnrtrmm commented:
Okay, I think there are at least the following three issues:

- Young forest area and carbon pools don't match. This should be fixed by either increasing initial young forest area or decreasing initial young forest carbon pools, depending on what the literature says. In any case, the young forest area and carbon pool fractions have to match (biomass and soil carbon fractions should probably be smaller than area fractions, as carbon pools are still recovering). Fixing this by increasing the initial young forest area is helpful with the terrestrial carbon balance. Decreasing the initial carbon pools has no major effect.
- The litter and soil carbon decay parameters should be looked at. Some of these parameters are different from those in a 2018 LPJmL version published in GMD (natural decay rate litter, natural decay rate fast soil carbon and temp response). When setting these parameters to the 2018 values, the terrestrial carbon balance looks much better.
- Currently the young forest area drops almost exponentially, which means we might have too little (re)forestation. The new calibration data I put in for forest regrowth show that the corresponding carbon uptake is rather increasing slightly between 1980 and 2022, meaning that young forest area should also increase a bit during this time.

wasbridge commented:
@jnnsbrr if you can adjust the parameter values that @lnnrtrmm is talkingn about, I can then teach you how to modify the machine calibration that I setup for food and land use to fix the young forest / degraded land amounts.  Just lemme know if/when you want that help.

jnnsbrr commented:
Dear @lnnrtrmm and @wasbridge,
* I fixed the young forest area and stock issue by including the share of young forest of the land use in the forest routine, that's clearly something I totally forgot to do - thanks @lnnrtrmm ! And introduced a Young mature forest biomass ratio instead to include the difference of above ground biomass in young and mature forests
* regarding the soil carbon decay, @lnnrtrmm I thought the LPJmL4 Parametrization make it worse? I talked to the person who implemented/changed that in LPJmL and she said we can use that as a sensitivity parameter, so both would be fine since the literature regarding these values is also quite diverse (have not looked into that though)
* I do not get your last point though, maybe it has changed now with the fixes?`
* Also I changed the crop residue on field fraction in the same way we did it for the harvest index. We have currently no demand for the crop residues for other use. So I changed it in a way that it more or less be constant (slight increase) - conservative assumptions. With the effect that more crop residues go into the litter -> soil in the future. Also I decreased the harvest index based on the literature with the same effect. These assumptions are based on literature values.

Major issue I found is the way we overestimate grass fed animals. Somehow I did not see that assumption earlier but we assume that only 33% of animals are crop fed. On the other hand only 10-15 percent of animals globally are rouminants (which eat grass). And only a fraction of that is actually fed by grass. So I introduced a max value 5% of grass fed animals for `animal_products_calorie_requirements_met_via_grazing`. [Here is some more information on that](https://oms-www.files.svdcdn.com/production/downloads/reports/fcrn_gnc_report.pdf) We are actually totally overestimating grazing.
This is probably not done in a nice way and maybe I am missing something! But it seems this changes our carbon source to a carbon sink.

lnnrtrmm commented:
Hi @jnnsbrr,

I had written to you on Mattermost regarding the soil decay parameters. For two of them going to LPJmL4 indeed makes it a bit worse, but the major effect comes from the third parameter, which is the natural litter decay rate. We currently have it at 0.7, whereas LPJmL has it at 0.3, meaning that we only put 30% of all litter carbon into the soil, whereas in LPJmL it is 70% (was there maybe a mix-up of the numbers?). This is the main parameter determining carbon input to the soil, so it has the dominating effect of the three parameters I mentioned.

My last point was actually only true for cases, where I had set the initial young forest area fraction to significantly more than the current 0.5%. It is still the case though that we need more reforestation to match the reforesting carbon uptake data I got from the GCB...

wasbridge commented:
@jnnsbrr on the 33% of crops for livestock - https://openknowledge.fao.org/server/api/core/bitstreams/c93da831-30b3-41dc-9e12-e1ae2963abde/content "33 percent of croplands are used for livestock feed production". I guess your point is that livestock isn't all animal products?  I can uncontrain this but what fraction of crops do you expect to be consumed in the process of creaeting animal products?

jnnsbrr commented:
I guess it could also be [a bit more](https://ourworldindata.org/global-land-for-agriculture#:~:text=38%25%20is%20for%20livestock%20feed), but you are right this is a bottleneck if we constrain both ...

wasbridge commented:
@jnnsbrr @lnnrtrmm Okay -- made what I think is significant progress.

1. UNIT GtC per PCal I assume is for crops (corn, wheat, soybean, yada yada yada) because that is where it was used initially.  I also assume that value was pulled from LPJML for those species.  My gut says that number shouldn't be the same for grass in a field that is being consumed right?  So that is what the GtC per Pcal of grazing multipler is taking into account.  I set it to 5% to reproduce what @jnnsbrr was wanting for Carbon from grazing... I have no idea if that is right or wrong. 

2. I added a new component to "yield" (grass net primary production) of grassland... which is the amount of grassland itself.  If grassland goes down, that means everything else being equal the yield of the average Hectacre of land should go up!  THe assumption being the worst most unproductive land is abandoned and when you abandon bad land, the average for all land goes up!

3. I recalibrated changes in land use types to nail the two land use change based balances re: deforestation and young forests.

4. I added a new calibration to set the NPP parameters in the grass and forest modules to get the terrestial carbon balance in line.

Doing all of this makes it match GCP data.  I will commit shortly

wasbridge commented:
Whoops! Got the grazing NPP effect backwards!  Once I put it the proper way it calibrated out to 0 -- I removed it!

wasbridge commented:
If grass net primary production in 1980 is adjustable we can calibrate bang on without issue 

See Run 2 - with grass net primary production in 1980 of 0.00575

![image](https://github.com/user-attachments/assets/d28ab887-afa6-4c95-8387-72b7fa5e4c2f)

lnnrtrmm commented:
Hi @wasbridge,

this looks very good already, also the match to young forest regrowth is really nice! I wonder whether some of the discussed soil and litter parameters could be used for calibration too, in case there is some unrealistic behaviour in the future @jannes? As Chris Smith just wrote in the dev climate impacts channel, we will have to match atmospheric CO2 calibration data very well and have the model produce reasonable fluxes also under idealized 1pct co2 scenarios.

For the moment, I think this could already be merged into FRIDA mainline? We can then do the fine tuning together with the ocean carbon cycle in a separate atmospheric CO2 stock to produce the right overall fluxes and atmospheric CO2 and then we can remove the FAIR component and connect the new stock instead (probably in a different pull request?).

wasbridge commented:
@lnnrtrmm lets discuss with @chrisdwells and Chris Smith once @jnnsbrr is comfortable with what is here both behaviorally and structurally :-).

jnnsbrr commented:
Dear @wasbridge, great thanks a lot this looks good already!
I still see some problems but I guess it would be better to discuss them directly. We could meet today if you have time? @chrisdwells, @lnnrtrmm and @wasbridge ?

jnnsbrr commented:
To collect some of the problems ..
* land cycling is off again ...
![image](https://github.com/user-attachments/assets/db7e6073-f29b-4f6c-906c-b4de6566556c) . Its hard to find good literature data for that but I would be rather conservative here and not let it be more then 1/3 of the extent of grassland.
* Degraded land is up again as well
* UNIT GtC per PCal - since it was missing a reference I was searching again (its not from LPJmL) we don't do calories. Found some and increased the value slightly. Also yes you are absolutely right with calorie content of grass, in general its quite high, but animals cannot digeste it well, rouminants are better but still miss a lot of calories though. Again its hard to find good literature value and I am not an expert on this ...
* yes grassland NPP is quite uncertain, still ours is quite high compared to literature values, so yes we can change it but not increase it much further

wasbridge commented:
@jnnsbrr @lnnrtrmm Confirm you're happy with the calibration and I think we're good to go here.

jnnsbrr commented:
Looks good! Ready to merge from my side.

lnnrtrmm commented:
Hi @wasbridge, I think it is generally nice what the land carbon cycle structure looks like now, and I don't want to be pedantic, but I would like to point out that the fit of the terrestrial carbon balance is still missing approximately 1 GtC per year at the end of the initialisation period. If we would connect this back to the atmosphere, we would be off by some 30-40 GT C in 2020. But we really have to have this match almost perfectly (except from the large natural variability in the calibration data) to be able to get the atmospheric CO2 concentration right by ~2 ppm, as Chris Smith requested it (and I agree that we should really be precise with atmospheric CO2 in our model).

![image](https://github.com/user-attachments/assets/46ecd40f-268c-4dc4-8710-f6f99d5e9c77)

Now, since this is not yet coupled to the atmospheric CO2 cycle, we can of course merge this into FRIDA mainline, and then do the final calibration only after I have put in the ocean carbon cycle (which I can probably do by the end of today). I would be happy to do the overall calibration.

Another thought: Maybe some of the misfit in the terrestrial carbon balance is still because the model also tries to fit the faulty calibration data for land-use co2 emissions in the emissions module?

jnnsbrr commented:
I found something which is - for some reason `CO2 tree net primary
production parameter` for forest is much smaller than for grass and crop (also the others for CO2 and Temperature effect parameters). It seems we can do a lot with these Parameters and match GCB quite well here ...

wasbridge commented:
@jnnsbrr We can work on this Friday maybe?
